File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        vector<vector<int>> ans;
        // set<vector<int>> st;
        sort(candidates.begin(),candidates.end());
        helper(temp,0,ans,candidates,target);
        
        return ans;
    }
    
    void helper(vector<int>& temp,int index,vector<vector<int>>& ans,vector<int>& candidates,int target){
        if(index == candidates.size()){
            if(target == 0){
                ans.push_back(temp);
            }   
            return;
        }
        
        
        if(candidates[index] <= target){
            temp.push_back(candidates[index]);
            helper(temp,index,ans,candidates,target-candidates[index]);
            temp.pop_back();
        }
        helper(temp,index+1,ans,candidates,target);
        
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.s = set()

        def dfs(arr,i,curSum):
            # if arr == [2, 2, 3] or arr==[7]:
            #     print(curSum,i)
            if curSum == target:
                self.s.add(tuple(arr))
                return
            if i == len(candidates) or curSum > target or candidates[i] > target or curSum + candidates[i] > target :
                return
            
            
            
            # same index again
            arr.append(candidates[i])
            dfs(arr,i,curSum + candidates[i])
            arr.pop() #needs some condition
            # next index
            dfs(arr,i+1,curSum)
        
        dfs([],0,0)

        return [list(i) for i in self.s]
        
